src/combinatorics/hall.lean

```diff
@@ -0,0 +1,382 @@
+/-
+Copyright (c) 2021 Alena Gusakov, Bhavik Mehta, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alena Gusakov, Bhavik Mehta, Kyle Miller
+-/
+import data.fintype.basic
+import data.rel
+import data.set.finite
+
+/-!
+# Hall's Marriage Theorem
+
+Given a list of finite subsets $$X_1,X_2,\dots,X_n$$ of some given set
+$$S$$, Hall in [Hall1935] gave a necessary and sufficient condition
+for there to be a list of distinct elements $$x_1,x_2,\dots,x_n$$ with
+$$x_i∈ X_i$$ for each $$i$$: it is when the union of any $$k$$ of
+these subsets has at least $$k$$ elements.
+
+This file proves this for an indexed family `s : ι → finset α` of
+finite sets, with `[fintype ι]`, along with some variants of the
+statement.  The list of distinct representatives is given by an
+injective function `f : ι → α` such that `∀ i, f i ∈ s i`.
+
+## Main statements
+* `all_card_le_bind_card_iff_exists_injective` is in terms of `s : ι → finset α`.
+* `all_card_le_rel_image_card_iff_exists_injective` is in terms of a relation
+  `r : α → β → Prop` such that `rel.image r {a}` is a finite set for all `a : α`.
+* `all_card_le_filter_rel_iff_exists_injective` is in terms of a relation
+  `r : α → β → Prop` on finite types, with the Hall condition given in terms of
+  `finset.univ.filter`.
+
+## Todo
+
+* The theorem is still true even if `ι` is not a finite type.  The infinite case
+  follows from a compactness argument.
+* The statement of the theorem in terms of bipartite graphs is in preparation.
+
+## Tags
+
+Hall's Marriage Theorem, indexed families
+-/
+
+open finset
+
+universes u v
+
+namespace hall_marriage_theorem
+
+lemma inj_of_dite_disjoint_inj {α β : Type*} (s : set α) [decidable_pred s]
+  (f : s → β) (f' : sᶜ → β)
+  (hf : function.injective f) (hf' : function.injective f')
+  (im_disj : ∀ {x x' : α} {hx : x ∈ s} {hx' : ¬ x' ∈ s}, f ⟨x, hx⟩ ≠ f' ⟨x', hx'⟩) :
+  function.injective (λ x, if h : x ∈ s then f ⟨x, h⟩ else f' ⟨x, h⟩) :=
+begin
```

Comment:
I didn't notice that there was some Zulip discussion, and I modified the statement a little so that it is in terms of sets and complements.  It's functionally equivalent to the original -- it's just a change in notation -- and I did it because it matches how this lemma is used (and because I can see what the relationship is between `f` and `f'` easier).

Reply:
I've put a question there about it.  Either form will work for the theorem (and they have defeq types).